### PR TITLE
Optional entity length for range requests

### PR DESCRIPTION
**Pull request by marcospereira**
## Pull Request Checklist

* [x] Have you checked that both Scala and Java APIs are updated?
* [x] Have you added tests for any changed functionality?

## Purpose

Support optional entity-length for range requests, as documented by [RFC 7233](http://tools.ietf.org/html/rfc7233#section-4.2). This PR can be safety backported to 2.5.x after being merged.

## References

See #5961.

----------------

@schmitch and @mkurz you may be interested in reviewing this. ;-)

### CODE REVIEW

File: framework/src/play/src/main/java/play/mvc/RangeResults.java [hunk 1]
```diff
@@ -34,6 +34,60 @@
...
+     */
+    public static Result ofStream(InputStream stream, long contentLength, String filename) {
+        return JavaRangeResult.ofStream(contentLength, stream, rangeHeader(), filename, Optional.empty());
+    }
+
+    /**
+     * Returns the stream as a result considering "Range" header. If the header is present and
+     * it is satisfiable, then a Result containing just the requested part will be returned.
+     * If the header is not present or is unsatisfiable, then a regular Result will be returned.
+     *
```
Comment (schmitch) [inline framework/src/play/src/main/java/play/mvc/RangeResults.java]: Maybe we should actually make a note that contentLength is nullable?

**Comment by schmitch** [comment]
lgtm (the only thing which could maybe be better is that the methods without the content length should have a note that these should only be used if the content length could not be determined, but that is just for RFC correctness and needs only be a small note inside a comment, so nothing trivial..)

**Comment by gmethvin** [comment]
@marcospereira LGTM. Can you resolve the conflicts and merge?

**Comment by marcospereira** [comment]
@gmethvin rebased and green build. ;-)

**Comment by gmethvin** [comment]
2.5.x: 17bf0f0

**Pull request merged**